toolbox/relnotes/main.go

```diff
@@ -0,0 +1,784 @@
+			log.Printf("error during copy to file: %s", err)
+			return
+		}
+		f.WriteString("\n")
+	} else {
+		log.Printf("Failed to find draft - creating generic template... (error message/status code printed below)")
+		if err != nil {
+			log.Printf("Error message: %s", err)
+		} else {
+			log.Printf("Response status code: %d", resp.StatusCode)
+		}
+		f.WriteString("## Major Themes\n\n* TBD\n\n## Other notable improvements\n\n* TBD\n\n## Known Issues\n\n* TBD\n\n## Provider-specific Notes\n\n* TBD\n\n")
+	}
+
+	// Aggregate all previous release in series
+	f.WriteString(fmt.Sprintf("### Previous Release Included in %s\n\n", release))
+
+	// Regexp Example:
+	// Assume the release tag is v1.7.0, this regexp matches "- [v1.7.0-" in
+	//     "- [v1.7.0-rc.1](#v170-rc1)"
+	//     "- [v1.7.0-beta.2](#v170-beta2)"
+	reAnchor, _ := regexp.Compile(fmt.Sprintf("- \\[%s-", release))
+
+	resp, err = http.Get(changelogURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	if err == nil && resp.StatusCode == 200 {
+		buf := new(bytes.Buffer)
+		buf.ReadFrom(resp.Body)
+		for _, line := range strings.Split(buf.String(), "\n") {
+			if anchor := reAnchor.FindStringSubmatch(line); anchor != nil {
+				f.WriteString(line + "\n")
+			}
+		}
+		f.WriteString("\n")
+	} else {
+		log.Printf("Failed to fetch past changelog for minor release - continuing... (error message/status code printed below)")
+		if err != nil {
+			log.Printf("Error message: %s", err)
+		} else {
+			log.Printf("Response status code: %d", resp.StatusCode)
+		}
+	}
+}
+
+// patchRelease performs a patch (vX.Y.Z) release by printing out all the related changes.
+func patchRelease(f *os.File, info *ReleaseInfo) {
+	// Release note for different labels
+	f.WriteString(fmt.Sprintf("## Changelog since %s\n\n", info.startTag))
+
+	if len(info.releaseActionRequiredPRs) > 0 {
+		f.WriteString("### Action Required\n\n")
+		for _, pr := range info.releaseActionRequiredPRs {
+			f.WriteString(fmt.Sprintf("* %s (#%d, @%s)\n", extractReleaseNoteFromPR(info.prMap[pr]), pr, *info.prMap[pr].User.Login))
+		}
+		f.WriteString("\n")
+	}
+
+	if len(info.releasePRs) > 0 {
+		f.WriteString("### Other notable changes\n\n")
+		for _, pr := range info.releasePRs {
+			f.WriteString(fmt.Sprintf("* %s (#%d, @%s)\n", extractReleaseNoteFromPR(info.prMap[pr]), pr, *info.prMap[pr].User.Login))
+		}
+		f.WriteString("\n")
+	} else {
+		f.WriteString("**No notable changes for this release**\n\n")
+	}
+}
+
+// extractReleaseNoteFromPR tries to fetch release note from PR body, otherwise uses PR title.
+func extractReleaseNoteFromPR(pr *github.Issue) string {
+	// Regexp Example:
+	// This regexp matches the release note section in Kubernetes pull request template:
+	// https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
+	re, _ := regexp.Compile("```release-note\r\n(.+)\r\n```")
+	if note := re.FindStringSubmatch(*pr.Body); note != nil {
+		return note[1]
+	}
+	return *pr.Title
```

Comment:
do we still have this case now? i.e., a PR has a `release-note` label but without `release-note` block in PR body? @cjwagner

Reply:
If the PR has the `release-note` label then there must be a release-note block. The only `release-note-*` label that does not require a release-note block is `release-note-none`.

Reply:
I find an example PR which has the `release-note` label but without a release-note block: https://github.com/kubernetes/kubernetes/pull/54007#event-1295604801
The author (he is a repo maintainer) manually added the `release-note` label...

So the code here is correct...